torch/csrc/distributed/rpc/process_group_agent.cpp

```diff
@@ -193,16 +193,7 @@ void ProcessGroupAgent::join() {
       lock, [this] { return futures_.empty() && futureTimeouts_.empty(); });
   lock.unlock();
   pg_->barrier()->wait();
-  int dst = (pg_->getRank() + 1) % pg_->getSize();
-  enqueueSend(
-      SendWork(allWorkerInfo_[dst], Message({}, {}, MessageType::SHUTDOWN)));
   threadPool_.waitWorkComplete();
-  listenerThread_.join();
-  rpcRunning_.store(false);
-  // This is needed in case no futures were created, otherwise the future
-  // timeout watchdog would sleep forever.
-  futureTimeoutCV_.notify_one();
-  futureTimeoutThread_.join();
   PythonRpcHandler::getInstance().cleanup();
```

Comment:
This should not be backend specific and should be moved to a common place. One way would be to have an implementation for `shutdown` in RpcAgent which calls this and then calls a pure virtual method `shutdownImpl`. Each backend can specify their own shutdown mechanism and the common parts are pulled into RpcAgent.

Reply:
We do call `_cleanup_python_rpc_handler()` in Python `wait_all_workers/join_rpc`. @zhaojuanmao Could you please confirm whether this is still necessary here?

Reply:
@mrshenli  and @rohan-varma _cleanup_python_rpc_handler() should be only called in python side and it should be called in both wait_all_workers/join_rpc and shutdown() API as these two APIs can be called independently, right? 

if uses write a pure C++ application, there is no python interpreter, we can not clean up in C++ in this case. I guess PythonRpcHandler::getInstance().cleanup() is added accidentally by some diff some time (I did not aware it was added in C++)?

Reply:
@zhaojuanmao thanks for clarifying. 

@rohan-varma As for now we require applications to call `wait_all_workers()` followed by `shutdown()`, it's probably OK that we just call it in `shutdown()`. But not necessary to trigger tests again due to this. We can discuss in a followup PR about whether we can remove this call in join.